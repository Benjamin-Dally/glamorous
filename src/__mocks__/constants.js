module.exports = {
  CHANNEL: '__glamorous__',
  isPreact: Boolean(process.env.MOCK_PREACT),
}
