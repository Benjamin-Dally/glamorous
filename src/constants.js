export const CHANNEL = '__glamorous__'
