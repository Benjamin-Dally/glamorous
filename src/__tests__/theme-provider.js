/* eslint func-style:0 */
import React, {Component} from 'react'
import {render, mount} from 'enzyme'
import glamorous from '../'
import ThemeProvider from '../theme-provider'
import {CHANNEL} from '../constants'

const getMockedContext = unsubscribe => ({
  [CHANNEL]: {
    getState: () => {},
    setState: () => {},
    subscribe: () => unsubscribe,
  },
})

test('renders a component with theme', () => {
  const Comp = glamorous.div(
    {
      color: 'red',
    },
    ({theme}) => ({padding: theme.padding}),
  )
  expect(
    render(
      <ThemeProvider theme={{padding: '10px'}}>
        <Comp />
      </ThemeProvider>,
    ),
  ).toMatchSnapshot()
})

test('theme properties updates get propagated down the tree', () => {
  class Parent extends Component {
    state = {
      padding: 10,
    }

    render() {
      return (
        <ThemeProvider theme={{padding: this.state.padding}}>
          <Child />
        </ThemeProvider>
      )
    }
  }
  const Child = glamorous.div(
    {
      color: 'red',
    },
    ({theme}) => ({padding: theme.padding}),
  )
  const wrapper = mount(<Parent />)
  expect(wrapper).toMatchSnapshot(`with theme prop of padding 10px`)
  wrapper.setState({padding: 20})
  expect(wrapper).toMatchSnapshot(`with theme prop of padding 20px`)
})

test('merges nested themes', () => {
  const One = glamorous.div({}, ({theme: {padding, margin}}) => ({
    padding,
    margin,
  }))
  const Two = glamorous.div({}, ({theme: {padding, margin}}) => ({
    padding,
    margin,
  }))
  expect(
    mount(
      <div>
        <ThemeProvider theme={{padding: 1, margin: 1}}>
          <div>
            <One />
            <ThemeProvider theme={{margin: 2}}>
              <Two />
            </ThemeProvider>
          </div>
        </ThemeProvider>
      </div>,
    ),
  ).toMatchSnapshot()
})

test('renders if children are null', () => {
  expect(
    mount(
      <ThemeProvider theme={{padding: 1}}>
        {false && <div />}
      </ThemeProvider>,
    ),
  ).toMatchSnapshot()
})

test('cleans up outer theme subscription when unmounts', () => {
  const unsubscribe = jest.fn()
  const context = getMockedContext(unsubscribe)
  const wrapper = mount(<ThemeProvider theme={{}} />, {context})
  wrapper.unmount()
  expect(unsubscribe).toHaveBeenCalled()
})

test('does nothing when receive same theme via props', () => {
  const theme = {margin: 2}
  const wrapper = mount(<ThemeProvider theme={theme} />)
  expect(wrapper).toMatchSnapshot(`with theme prop of margin 2px`)
  wrapper.setProps({theme})
  expect(wrapper).toMatchSnapshot(`with theme prop of margin 2px`)
})
