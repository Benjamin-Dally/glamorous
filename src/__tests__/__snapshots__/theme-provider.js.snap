// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does nothing when receive same theme via props: with theme prop of margin 2px 1`] = `
<ThemeProvider
  theme={
    Object {
      "margin": 2,
    }
  }
/>
`;

exports[`does nothing when receive same theme via props: with theme prop of margin 2px 2`] = `
<ThemeProvider
  theme={
    Object {
      "margin": 2,
    }
  }
/>
`;

exports[`merges nested themes 1`] = `
.glamor-0,
[data-glamor-0] {
  padding: 1px;
  margin: 1px;
}

.glamor-1,
[data-glamor-1] {
  padding: 1px;
  margin: 2px;
}

<div>
  <ThemeProvider
    theme={
      Object {
        "margin": 1,
        "padding": 1,
      }
    }
  >
    <div>
      <glamorous(div)>
        <div
          className="glamor-0"
        />
      </glamorous(div)>
      <ThemeProvider
        theme={
          Object {
            "margin": 2,
          }
        }
      >
        <glamorous(div)>
          <div
            className="glamor-1"
          />
        </glamorous(div)>
      </ThemeProvider>
    </div>
  </ThemeProvider>
</div>
`;

exports[`propagates theme updates through nested ThemeProviders 1`] = `
.glamor-0,
[data-glamor-0] {
  background-color: black;
  color: red;
}

<ThemeProvider
  theme={
    Object {
      "bg": "black",
    }
  }
>
  <ThemeProvider
    theme={[Function]}
  >
    <glamorous(div)>
      <div
        className="glamor-0"
      />
    </glamorous(div)>
  </ThemeProvider>
</ThemeProvider>
`;

exports[`renders a component with theme 1`] = `
.glamor-0,
[data-glamor-0] {
  color: red;
  padding: 10px;
}

<div
  class="glamor-0"
/>
`;

exports[`renders if children are null 1`] = `
<ThemeProvider
  theme={
    Object {
      "padding": 1,
    }
  }
/>
`;

exports[`renders inverted theme 1`] = `
.glamor-0,
[data-glamor-0] {
  background-color: white;
  color: palevioletred;
}

.glamor-1,
[data-glamor-1] {
  background-color: palevioletred;
  color: white;
}

<div>
  <div
    class="glamor-0"
  />
  <div
    class="glamor-1"
  />
</div>
`;

exports[`theme properties updates get propagated down the tree: with theme prop of padding 10px 1`] = `
.glamor-0,
[data-glamor-0] {
  color: red;
  padding: 10px;
}

<Component>
  <ThemeProvider
    theme={
      Object {
        "padding": 10,
      }
    }
  >
    <glamorous(div)>
      <div
        className="glamor-0"
      />
    </glamorous(div)>
  </ThemeProvider>
</Component>
`;

exports[`theme properties updates get propagated down the tree: with theme prop of padding 20px 1`] = `
.glamor-0,
[data-glamor-0] {
  color: red;
  padding: 20px;
}

<Component>
  <ThemeProvider
    theme={
      Object {
        "padding": 20,
      }
    }
  >
    <glamorous(div)>
      <div
        className="glamor-0"
      />
    </glamorous(div)>
  </ThemeProvider>
</Component>
`;

exports[`throws if inverted theme is not object 1`] = `"[ThemeProvider] Please return an object from your theme function, i.e. theme={() => ({})}!"`;
